In [ ]:
import pandas as pd

total_unique_ids = set()

# Loops through 25 raw eye-tracking CSV files
for i in range(1, 26):
    df = pd.read_csv(f"../data/rawdata/{i}.csv")
    unique_ids = df["Participant"].unique()
    print(f"File {i}: Number of unique participants: {len(unique_ids)}")
    total_unique_ids.update(unique_ids)

print(f"Total unique participants across all files: {len(total_unique_ids)}")

File 1: Number of unique participants: 2
File 2: Number of unique participants: 1
File 3: Number of unique participants: 1
File 4: Number of unique participants: 2
File 5: Number of unique participants: 1
File 6: Number of unique participants: 2
File 7: Number of unique participants: 1


/var/folders/g3/x3j647bn42j_bpmd_9vl047w0000gn/T/ipykernel_78960/1338503006.py:6: DtypeWarning: Columns (13,14,15,16,17,22,23,24,25,26,27,28,29,30) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(f"../data/rawdata/{i}.csv")


File 8: Number of unique participants: 1
File 9: Number of unique participants: 1
File 10: Number of unique participants: 2


/var/folders/g3/x3j647bn42j_bpmd_9vl047w0000gn/T/ipykernel_78960/1338503006.py:6: DtypeWarning: Columns (15,16,17,18,19,20,21,22,23,24,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(f"../data/rawdata/{i}.csv")


File 11: Number of unique participants: 1
File 12: Number of unique participants: 1
File 13: Number of unique participants: 3
File 14: Number of unique participants: 5
File 15: Number of unique participants: 1
File 16: Number of unique participants: 2
File 17: Number of unique participants: 1
File 18: Number of unique participants: 1
File 19: Number of unique participants: 1
File 20: Number of unique participants: 1
File 21: Number of unique participants: 1
File 22: Number of unique participants: 2
File 23: Number of unique participants: 1
File 24: Number of unique participants: 2
File 25: Number of unique participants: 2
Total unique participants across all files: 10


shows only then unique participant IDs found in all files. But this is not true.

In [11]:
df = pd.read_csv("../data/1.csv")
print(df.shape)

(34630, 37)


/var/folders/g3/x3j647bn42j_bpmd_9vl047w0000gn/T/ipykernel_78960/553178503.py:1: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("../data/1.csv")


The file ../data/1.csv contains: 34,630 rows, 37 columns
	•	37 columns

In [ ]:

# ---------------------------------------------------------------
# Load all 25 raw CSV files from ../data/ and combine them into
# a single large DataFrame ("master").
#
# Each file contains raw eye-tracking samples exported by the
# EyeLink system. The files are not cleaned and may have different
# column sets, so pd.concat() keeps all columns that appear in any
# file (resulting in ~59 columns total).
#
# After concatenation, 'master' contains all raw samples across
# the dataset. This is useful for inspecting global structure,
# counting total rows, checking participant IDs, and examining
# the full feature space before preprocessing.
#
#   The output (2,252,943 rows × 59 columns) represents the
#   complete dataset before cleaning or splitting by participant.
#   This DataFrame is too large for feature extraction or modeling
#   and is only used for diagnostics and validation.
# ---------------------------------------------------------------

dfs = []

for i in range(1, 26):
    df = pd.read_csv(f"../data/{i}.csv")
    dfs.append(df)

master = pd.concat(dfs, ignore_index=True)
print(master.shape, "rows, cols ->", master.shape)

/var/folders/g3/x3j647bn42j_bpmd_9vl047w0000gn/T/ipykernel_78960/3998412500.py:4: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(f"../data/{i}.csv")
/var/folders/g3/x3j647bn42j_bpmd_9vl047w0000gn/T/ipykernel_78960/3998412500.py:4: DtypeWarning: Columns (7,15,16,17,18,19,20,21,22,23,24,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(f"../data/{i}.csv")
/var/folders/g3/x3j647bn42j_bpmd_9vl047w0000gn/T/ipykernel_78960/3998412500.py:4: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(f"../data/{i}.csv")
/var/folders/g3/x3j647bn42j_bpmd_9vl047w0000gn/T/ipykernel_78960/3998412500.py:4: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(f"../data/{i}.csv")
/var/folders/g3/x3j647bn42j_bpmd_9vl04

(2252943, 59) rows, cols -> (2252943, 59)


In [13]:
master.head(1).to_string()

"   Unnamed: 0  RecordingTime [ms] Time of Day [h:m:s:ms]     Trial                             Stimulus  Export Start Trial Time [ms]  Export End Trial Time [ms] Participant  Color  Tracking Ratio [%] Category Group Category Right Category Left Index Right Index Left Pupil Diameter Right [mm] Pupil Diameter Left [mm] Point of Regard Right X [px] Point of Regard Right Y [px] Point of Regard Left X [px] Point of Regard Left Y [px] AOI Name Right AOI Name Left Gaze Vector Right X Gaze Vector Right Y Gaze Vector Right Z Gaze Vector Left X Gaze Vector Left Y Gaze Vector Left Z Annotation Name Annotation Description Annotation Tags Mouse Position X [px] Mouse Position Y [px] Scroll Direction X Scroll Direction Y                           Content Pupil Size Right X [px] Pupil Size Right Y [px] Pupil Size Left X [px] Pupil Size Left Y [px] Eye Position Right X [mm] Eye Position Right Y [mm] Eye Position Right Z [mm] Eye Position Left X [mm] Eye Position Left Y [mm] Eye Position Left Z [mm] Pu

In [8]:
print(f"Total unique participants across all files: {len(total_unique_ids)}")
print(f"Unique Participants IDs: {total_unique_ids}")

Total unique participants across all files: 97
Unique Participants IDs: {'51', '23', 7, np.int64(11), 14, 15, 17, 18, np.int64(19), np.int64(20), 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 35, 37, 38, 39, np.int64(41), 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, '28', 55, 54, 58, '8', 59, '18', '3', '10', '14', '11', '45', '29', '27', '57', '40', '42', '6', '33', '22', '19', '35', '48', '9', '5', '54', '26', '34', '15', '4', '32', '39', '17', '25', '53', '13', '43', '59', '56', '2', '44', '7', '52', '50', '58', '49', '38', '21', '37', '55', '24', 'Unidentified(Neg)', '31', '36', '46', 'Unidentified(Pos)', '30', '20', '1', '47'}


The dataset contains **97 unique participant IDs** across all raw files.  
The IDs appear in mixed formats (strings, integers, `np.int64`) because the
`Participant` column is inconsistently typed across files. Some IDs are stored
as numbers (e.g., `7`, `15`), while others appear as strings (e.g., `'28'`, `'8'`)
or special labels such as `'Unidentified(Pos)'` and `'Unidentified(Neg)'`.

This mixed typing is expected in raw EyeLink exports and will be normalized
later during cleaning. The count of 97 reflects *all distinct labels* present
before unifying their types.

In [49]:
unique_ids = df["Participant"].unique()
print(f"Unique Participants IDs: {unique_ids}")
print("Number of unique participants:", len(unique_ids))

Unique Participants IDs: ['Unidentified(Neg)' 'Unidentified(Pos)']
Number of unique participants: 2


In [16]:
import pandas as pd
import re
from pathlib import Path
import csv

def normalize_name(name: str) -> str:
    # Remove BOM/whitespace, unify spaces/underscores, drop units in brackets into suffix-less form
    name = name.strip()
    # Remove bracketed units, keep a hint if helpful
    # e.g., "RecordingTime [ms]" -> "RecordingTime"
    name = re.sub(r"\[[^\]]*\]", "", name)
    # Collapse whitespace and punctuation to underscores
    name = re.sub(r"[^\w]+", "_", name)
    name = name.strip("_").lower()
    return name

# Map verbose original labels to a compact canonical schema
CANONICAL_MAP = {
    # basics
    "recordingtime": "recording_time_ms",
    "recordingtime_ms": "recording_time_ms",
    "time_of_day_h_m_s_ms": "time_of_day",
    "trial": "trial",
    "stimulus": "stimulus",
    "export_start_trial_time": "export_start_ms",
    "export_start_trial_time_ms": "export_start_ms",
    "export_end_trial_time": "export_end_ms",
    "export_end_trial_time_ms": "export_end_ms",
    "participant": "participant",
    "tracking_ratio": "tracking_ratio_pct",
    "tracking_ratio_pct": "tracking_ratio_pct",
    "color": "color",

    # pupil diameters
    "pupil_diameter_right_mm": "pupil_r_mm",
    "pupil_diameter_left_mm": "pupil_l_mm",

    # point of regard (pixels)
    "point_of_regard_right_x_px": "por_r_x_px",
    "point_of_regard_right_y_px": "por_r_y_px",
    "point_of_regard_left_x_px": "por_l_x_px",
    "point_of_regard_left_y_px": "por_l_y_px",
}

# Obvious noise to drop (you can adjust this list)
DROP_PATTERNS = [
    r"^unnamed",                # Unnamed: 0, etc.
    r"^mouse_",                 # Mouse Position
    r"^scroll_",                # Scroll
    r"^port_status$",           # Port
    r"^content$",               # Content (verbatim filename already in stimulus)
    r"^gaze_vector_",           # 3D gaze vectors (if unused)
    r"^eye_position_",          # Eye position in mm (if unused)
    r"^pupil_position_",        # Pupil position in px (if unused)
    r"^aoi_",                   # AOI metadata (if not needed now)
    r"^annotation_",            # Annotation Name/Description/Tags
    r"^binocular$",             # stray label
    r"^groupe_d_enfants$",      # localized extra
    r"^category", r"^index_",   # right/left category/index columns
]

def should_drop(col: str) -> bool:
    return any(re.match(pat, col) for pat in DROP_PATTERNS)

def load_and_clean_csv(path) -> pd.DataFrame:
    # Let pandas sniff comma vs semicolon; handle BOM
    df = pd.read_csv(path, sep=None, engine="python", encoding="utf-8-sig", quoting=csv.QUOTE_MINIMAL)

    # Normalize headers
    norm_cols = [normalize_name(c) for c in df.columns]
    df.columns = norm_cols

    # Drop obvious junk (Unnamed, mouse, etc.)
    keep_cols = [c for c in df.columns if not should_drop(c)]
    df = df[keep_cols]

    # Canonical rename (best-effort)
    rename_map = {c: CANONICAL_MAP.get(c, c) for c in df.columns}
    df = df.rename(columns=rename_map)

    # Add source for traceability
    df["source_file"] = Path(path).name

    # Parse/convert key types
    if "recording_time_ms" in df.columns:
        # Sometimes these are floats; keep numeric
        df["recording_time_ms"] = pd.to_numeric(df["recording_time_ms"], errors="coerce")

    if "export_start_ms" in df.columns:
        df["export_start_ms"] = pd.to_numeric(df["export_start_ms"], errors="coerce")
    if "export_end_ms" in df.columns:
        df["export_end_ms"] = pd.to_numeric(df["export_end_ms"], errors="coerce")

    for col in ["pupil_r_mm", "pupil_l_mm", "por_r_x_px", "por_r_y_px", "por_l_x_px", "por_l_y_px", "tracking_ratio_pct"]:
        if col in df.columns:
            df[col] = pd.to_numeric(df[col], errors="coerce")

    # Participant as string ID
    if "participant" in df.columns:
        df["participant"] = df["participant"].astype(str).str.strip()

    # Tidy: remove fully empty columns that may remain after coercion
    df = df.dropna(axis=1, how="all")

    return df

# ---- Use on one file
# df = load_and_clean_csv("data/raw/your_file.csv")
# df.head()

# ---- Or: batch all files into a master
def build_master(folder="data/raw"):
    paths = sorted(Path(folder).glob("*.csv"))
    dfs = [load_and_clean_csv(p) for p in paths]
    master = pd.concat(dfs, ignore_index=True, sort=False)

    # Optional: keep only the canonical subset for analysis
    canonical_subset = [
        "source_file",
        "participant", "trial", "stimulus", "color",
        "recording_time_ms", "time_of_day",
        "export_start_ms", "export_end_ms",
        "tracking_ratio_pct",
        "pupil_r_mm", "pupil_l_mm",
        "por_r_x_px", "por_r_y_px", "por_l_x_px", "por_l_y_px",
    ]
    have = [c for c in canonical_subset if c in master.columns]
    master = master[have]

    return master

# Example:
# master = build_master("data/raw")
# master.to_parquet("data/processed/master.parquet", index=False)
# master.to_csv("data/processed/master.csv", index=False)

In [17]:
master = build_master("../data/")

In [21]:
print(master.head(5).to_string())
print(master.shape)

  source_file participant     trial                             stimulus  color  recording_time_ms   time_of_day  export_start_ms  export_end_ms  tracking_ratio_pct
0       1.csv          50  Trial020  20 eye tracking (ballon droite).avi  Coral        4780015.315  13:22:17:254              0.0      52058.676             78.5652
1       1.csv          50  Trial020  20 eye tracking (ballon droite).avi  Coral        4780020.865  13:22:17:260              0.0      52058.676             78.5652
2       1.csv          50  Trial020  20 eye tracking (ballon droite).avi  Coral        4780040.758  13:22:17:279              0.0      52058.676             78.5652
3       1.csv          50  Trial020  20 eye tracking (ballon droite).avi  Coral        4780080.503  13:22:17:319              0.0      52058.676             78.5652
4       1.csv          50  Trial020  20 eye tracking (ballon droite).avi  Coral        4780120.360  13:22:17:359              0.0      52058.676             78.5652
(2253002, 

In [27]:
total_unique_ids = master["participant"].unique()

print(f"Total unique participants across all files: {len(total_unique_ids)}")
print(f"Unique Participants IDs: {total_unique_ids}")

Total unique participants across all files: 60
Unique Participants IDs: ['50' '52' '49' '55' '53' '46' '42' 'Unidentified(Neg)'
 'Unidentified(Pos)' '20' '19' '27' '17' '15' '21' '23' '28' '18' '43'
 '48' '51' '44' '45' '47' '13' '14' '54' '59' '25' '26' '58' '29' '56'
 '30' '57' '31' '36' '2' '3' '4' '5' '6' '1' '9' '10' '8' '34' '40' '38'
 '32' '35' '11' '33' '39' '37' '7' '41' '22' '24' nan]


In [28]:
numeric = pd.to_numeric(total_unique_ids, errors='coerce')

# 2️⃣ Build a DataFrame to keep both
df_ids = pd.DataFrame({'raw': total_unique_ids, 'num': numeric})

# 3️⃣ Sort first by 'num' (numeric ascending), then by raw (so text labels come after numbers)
df_ids_sorted = df_ids.sort_values(
    by=['num', 'raw'],
    key=lambda col: col if col.name != 'num' else col.fillna(1e9)  # big number for NaNs
)

# 4️⃣ Rebuild a clean, ordered list
ordered_ids = df_ids_sorted['raw'].dropna().tolist()

print("Ordered IDs:")
print(ordered_ids)

Ordered IDs:
['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '13', '14', '15', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '50', '51', '52', '53', '54', '55', '56', '57', '58', '59', 'Unidentified(Neg)', 'Unidentified(Pos)']


In [46]:
# prefer using the notebook's load_and_clean_csv (defined in an earlier cell)
lookFor = "2"
matches = []

for i in range(1, 26):
    path = f"../data/{i}.csv"

    # try to use the canonical loader if available, fall back to raw read
    try:
        df_i = load_and_clean_csv(path)
    except NameError:
        # load_and_clean_csv not defined in session; do a tolerant read
        df_i = pd.read_csv(path, sep=None, engine="python", encoding="utf-8-sig")
    except Exception as e:
        print(f"Failed to read {path}: {e}")
        continue

    # find participant column robustly (normalized loader will provide 'participant')
    if "participant" in df_i.columns:
        part_col = "participant"
    else:
        # try case-insensitive match to common original header
        cols_lower = {c.lower(): c for c in df_i.columns}
        part_col = cols_lower.get("participant")



    subset = df_i[df_i[part_col] == lookFor]
    if not subset.empty:
        print(f"File {i} ({path}) has {len(subset)} rows with participant == {lookFor}")
        matches.append(subset)

if matches:
    all_matches = pd.concat(matches, ignore_index=True, sort=False)
    print("Total rows with participant == 'Unidentified(Neg)':", len(all_matches))
else:
    print("No matches found for", lookFor)

File 9 (../data/9.csv) has 1500 rows with participant == 2
File 12 (../data/12.csv) has 2187 rows with participant == 2
File 18 (../data/18.csv) has 1613 rows with participant == 2
File 19 (../data/19.csv) has 1832 rows with participant == 2
Total rows with participant == 'Unidentified(Neg)': 7132


In [47]:
print("Matches details:")
for i, match in enumerate(matches, start=1):
    print(f"Match {i}:")
    print(match.head(1).to_string()) 

Matches details:
Match 1:
      recording_time_ms   time_of_day     trial       stimulus  export_start_ms  export_end_ms participant color  tracking_ratio_pct point_of_regard_right_x point_of_regard_right_y point_of_regard_left_x point_of_regard_left_y source_file
6702         488208.202  07:51:33:379  Trial019  FEDE Drte.avi              0.0      48854.563           2  Cyan             29.9665                       -                       -                      -                      -       9.csv
Match 2:
      recording_time_ms   time_of_day     trial                              stimulus  export_start_ms  export_end_ms participant    color  tracking_ratio_pct pupil_size_right_x pupil_size_right_y pupil_diameter_right pupil_size_left_x pupil_size_left_y pupil_diameter_left point_of_regard_right_x point_of_regard_right_y point_of_regard_left_x point_of_regard_left_y source_file
4141        2097233.054  08:17:10:000  Trial001  Federica Final_WMV_3000Kbps_720p.avi              0.0     